mzLib/MassSpectrometry/MzSpectra/SpectralSimilarity.cs

```diff
@@ -344,18 +405,48 @@ private double[] NormalizeSpectrumSum(double[] spectrum)
             double similarityScore = 1 - (2 * combinedEntropy - theoreticalEntropy - experimentalEntropy) / Math.Log(4);
             return similarityScore;
         }
-      
-        
-        public double? KullbackLeiblerDivergence_P_Q()
+
+        /// <summary>
+        /// This is used to calculate the KullbackLeibler divergence between a theoretical and experimental isotopic envelope
+        /// When using, the allPeaks argument in the SpectralSimilarity constructor should be set to "true"
+        /// </summary>
+        /// <param name="correctionConstant"> A constant value added to each intensity pair in order to penalize zero peaks.
+        /// Default is 1e-9 (one OoM smalller than the theoretical isotopic intensity minimum)</param>
+        /// <returns> A nullable double between 0 and positive infinity. More similar envelopes score lower </returns>
+        public double? KullbackLeiblerDivergence_P_Q(double correctionConstant = 1e-9)
         {
+            // counts the number of zero values. If zeroCount == intensityPairs.Count, then there are no shared peaks
+            int zeroCount = intensityPairs.Count(p => p.Item1 == 0 | p.Item2 == 0);
+            if (zeroCount == intensityPairs.Count) return null;
             double divergence = 0;
-            foreach (var pair in intensityPairs)
+
+            if (zeroCount == 0) // | correctionConstant == 0)
             {
-                if(pair.Item1 != 0 && pair.Item2 != 0)
+                foreach (var pair in intensityPairs)
                 {
-                    divergence += pair.Item1 * Math.Log(pair.Item1 / pair.Item2);
-                } 
+                    if (pair.Item1 != 0 && pair.Item2 != 0)
+                    {
+                        divergence += pair.Item1 * Math.Log(pair.Item1 / pair.Item2);
+                    }
+                }
             }
+            else
+            {
+                // Add correctionConstant and renormalize
+                // Need to use temp variables to avoid modifiying the Y array fields
+                double[] tempExperimentalYArray = Normalize(ExperimentalYArray.Select(i => i + correctionConstant).ToArray(), SpectrumNormalizationScheme.spectrumSum);
+                double[] tempTheoreticalYArray = Normalize(TheoreticalYArray.Select(i => i + correctionConstant).ToArray(), SpectrumNormalizationScheme.spectrumSum);
+                List<(double, double)> correctedIntensityPairs = CorrectedIntensityPairs(tempExperimentalYArray, tempTheoreticalYArray, allPeaks: true);
```

Comment:
can we just call the original intensity pairs function?

Reply:
Original intensity pairs function wouldn't take any arguments, instead pulling from the object's fields. I've edited the original function to have optional arguments.